src/integrator/Adress.cpp

```diff
@@ -413,95 +368,141 @@ namespace espressopp {
+            }
+            mayberoot = (lroot ? lroot : 0);
+            boost::mpi::all_reduce(*getSystem()->comm,mayberoot,root,boost::mpi::maximum<int>());
+            mpi::broadcast(*getSystem()->comm,verletList->adrPositions,root);
+
           }
-          mayberoot = (lroot ? lroot : 0);
-          boost::mpi::all_reduce(*getSystem()->comm,mayberoot,root,boost::mpi::maximum<int>());
-          mpi::broadcast(*getSystem()->comm,verletList->adrPositions,root); // only necessary for moving adrCenter
-       }
+          else if(verletList->getAdrList().size() > 1){
+
+            // New version (works also for many moving regions but is slower)
+            System& system = getSystemRef();
+            std::vector<Real3D*> procAdrPositions;
+            CellList realcells = getSystem()->storage->getRealCells();
+
+            verletList->adrPositions.clear();
+            for (CellListIterator it(realcells); it.isValid(); ++it) {
+
+                if (verletList->getAdrList().count(it->id()) == 1) {
+
+                   // Update the copy and append to adrPositions
+                   adrpos = it->position();
+		        ptr = &adrpos;
+		        procAdrPositions.push_back(ptr);
+
+		        //procAdrPositions.push_back(&(it->position())); // without the additional copy
+                }
+
+            }
+
+            std::vector<std::vector<Real3D*> > procAdrPositionsAll;
+            boost::mpi::all_gather(*system.comm, procAdrPositions, procAdrPositionsAll);
+
+            for (std::vector<std::vector<Real3D*> >::iterator itr=procAdrPositionsAll.begin(); itr != procAdrPositionsAll.end(); ++itr) {
+                for (std::vector<Real3D*>::iterator itr2=(*itr).begin(); itr2 != (*itr).end(); ++itr2){
+                  verletList->adrPositions.push_back(*itr2);
+                }
+            }
+
+          }
+          else{
+              std::cout << "When using moving AdResS regions there can be only either one or several of them. It seems like the list of PIDs of region defining particles is empty for some reason.\n";
+              exit(1);
+              return;
+          }
+
+        }
+
+        updatecount += 1;
+        if(updatecount == regionupdates){updatecount = 0;}
+
+      }
+
     }
-    
-    
-    
+
+
+
     // AdResS Weighting function
     real Adress::weight(real distanceSqr){
         if (dex2 > distanceSqr) return 1.0;
         else if (dexdhy2 < distanceSqr) return 0.0;
         else {
             real argument = sqrt(distanceSqr) - dex;
-            return 1.0-(30.0/(pow(dhy, 5.0)))*(1.0/5.0*pow(argument, 5.0)-dhy/2.0*pow(argument, 4.0)+1.0/3.0*pow(argument, 3.0)*dhy*dhy);
-            //return pow(cos(pidhy2 * argument),2.0); // for cosine squared weighting function
+            //return 1.0-(30.0/(pow(dhy, 5.0)))*(1.0/5.0*pow(argument, 5.0)-dhy/2.0*pow(argument, 4.0)+1.0/3.0*pow(argument, 3.0)*dhy*dhy);
+            return pow(cos(pidhy2 * argument),2.0); // for cosine squared weighting function
```

Comment:
Can you make the same change in VerletListAdressInteractionTemplate.hpp? It occurs twice, just search for cos, the commented lines are already there.